File: superstruct/sketches/bloom.py
"""Bloom filter sketch.

A bit array of m bits and k independent hash functions. To insert a
value we hash it k ways and set those bits. To check membership we
hash the same way and read those bits. If any bit is zero the value
was definitely never inserted. If all bits are one the value is
probably inserted with a tunable false positive rate.

We do not implement deletion. Standard bloom filters cannot delete
without tracking counters which would double the memory cost. The
false positive rate slowly rises as records churn through the sketch
over time. For a research demo this trade is fine.
"""
import hashlib
from typing import Any


class BloomSketch:
    """Probabilistic membership filter.

    Tiny memory footprint. False positives possible. False negatives
    never. Ideal as a fast gate in front of an expensive lookup.
    """

    def __init__(self, bit_size: int = 1 << 14, num_hashes: int = 5):
        # Default 16384 bits is 2 KB which is plenty for tens of
        # thousands of distinct values at a few percent false positive
        # rate. The user can dial this up for larger universes.
        self._bit_size = bit_size
        self._num_hashes = num_hashes

        # bytearray of bit_size / 8 bytes. All zeros initially.
        self._bits = bytearray(bit_size // 8)

    def _positions(self, value: Any) -> list[int]:
        """Compute the k bit positions for a given value.

        We hash repr(value) with md5 then chop the digest into k slices
        of four bytes each, taking each slice as a bit position modulo
        the filter size. Slow compared to a hand tuned bloom but works
        for any value the user might insert.
        """
        encoded = repr(value).encode("utf-8")
        digest = hashlib.md5(encoded).digest()
        # Double the digest so we can take overlapping windows when k
        # is larger than the digest can support naturally.
        doubled = digest + digest

        positions = []
        for i in range(self._num_hashes):
            start = (i * 4) % len(digest)
            chunk = doubled[start:start + 4]
            n = int.from_bytes(chunk, "big")
            positions.append(n % self._bit_size)
        return positions

    def add(self, value: Any) -> None:
        """Mark value as present in the filter."""
        for p in self._positions(value):
            self._bits[p // 8] |= (1 << (p % 8))

    def maybe_contains(self, value: Any) -> bool:
        """Return False if value was definitely never added.

        Return True when value was probably added with a small false
        positive rate. Never returns False on a value that was added.
        """
        for p in self._positions(value):
            if not (self._bits[p // 8] & (1 << (p % 8))):
                return False
        return True

File: src/sketch/bloom.rs
use crate::Value;
use md5::{Digest, Md5};

pub struct BloomSketch {
    bit_size: usize,
    num_hashes: usize,
    bits: Vec<u8>,
}

impl BloomSketch {
    pub fn new(bit_size: usize, num_hashes: usize) -> Self {
        Self {
            bit_size,
            num_hashes,
            bits: vec![0u8; bit_size / 8],
        }
    }

    pub fn default_size() -> Self {
        Self::new(1 << 14, 5)
    }

    fn positions(&self, value: &Value) -> Vec<usize> {
        let encoded = value.to_string();
        let digest = Md5::digest(encoded.as_bytes());
        let digest_bytes = digest.as_slice();
        let doubled: Vec<u8> = digest_bytes.iter().chain(digest_bytes.iter()).copied().collect();

        (0..self.num_hashes)
            .map(|i| {
                let start = (i * 4) % digest_bytes.len();
                let chunk = &doubled[start..start + 4];
                let n = u32::from_be_bytes(chunk.try_into().unwrap()) as usize;
                n % self.bit_size
            })
            .collect()
    }

    pub fn add(&mut self, value: &Value) {
        for p in self.positions(value) {
            self.bits[p / 8] |= 1 << (p % 8);
        }
    }

    pub fn maybe_contains(&self, value: &Value) -> bool {
        for p in self.positions(value) {
            if self.bits[p / 8] & (1 << (p % 8)) == 0 {
                return false;
            }
        }
        true
    }
}

impl Default for BloomSketch {
    fn default() -> Self {
        Self::default_size()
    }
}
